### PR TITLE
Add hexdigest verification to download

### PR DESCRIPTION
**Pull request by mberr**
This PR adds an optional file integrity check to the `download` method, where a user can provide a dictionary with hash algorithm to hexdigest pairs. If these digests are present, the method will verify matching digests of an existing file, and trigger a re-download if the sum does not match.

### CODE REVIEW

File: src/pystow/utils.py [hunk 1]
```diff
@@ -27,12 +28,85 @@
...
+    """
+    Check a file for hash sums.
+
+    :param destination:
+        The file path.
+    :param chunk_size:
+        The chunk size for reading the file.
+    :param hexdigests:
+        The expected hexdigests as (algorithm_name, expected_hex_digest) pairs.
+    :param verbose:
```
Comment (cthoyt) [inline src/pystow/utils.py:62]: why verbose? can't this just be handled by the user choosing the logging level?

File: src/pystow/utils.py [hunk 2]
```diff
@@ -48,8 +122,15 @@ def download(
...
     :raises KeyboardInterrupt: If a keyboard interrupt is thrown during download
     :raises ValueError: If an invalid backend is chosen
+    :raises HexDigestError: If the downloaded file's hexdigest does not match
     """
+    # input normalization
+    path = Path(path).resolve()
+
+    skip_download = False
     if os.path.exists(path) and not force:
+        skip_download = not get_offending_hexdigests(destination=path, hexdigests=hexdigests)
```
Comment (cthoyt) [inline src/pystow/utils.py:132]: I think we need to discuss what should happen if it's got a mismatch. Should it re-download, or should it throw an exception?

File: src/pystow/utils.py (hunk 1 above)
Comment (mberr) [reply src/pystow/utils.py:62]: :+1:

File: src/pystow/utils.py (hunk 2 above)
Comment (mberr) [reply src/pystow/utils.py:129]: I guess we can just fail fast. The user can provide `force=True` to get a fresh version of the file.

File: src/pystow/utils.py (hunk 1 above)
Comment (cthoyt) [reply src/pystow/utils.py]: Good then please remove the verbose argument

**Comment by cthoyt** [comment]
let's also add this check to the google download function while we're here

**Comment by mberr** [comment]
> let's also add this check to the google download function while we're here

It seems like the code organization for Google Drive download is a bit different:

For `download`, we check whether the path exists in the function. For the GDrive download, this is only done in the Module, but not in `download_from_google`.

File: src/pystow/utils.py [hunk 3]
```diff
@@ -306,35 +305,31 @@ def download_from_google(
...
     """
-    # TODO: Reduce code duplication with download
-    # input normalization
     path = Path(path).resolve()
 
-    if os.path.exists(path) and not force:
+    if path.exists() and not force:
         raise_on_digest_mismatch(path=path, hexdigests=hexdigests)
-        logger.debug('did not re-download %s for ID %s', path, file_id)
+        logger.debug('did not re-download %s from %s', path, file_id)
```
Comment (mberr) [inline src/pystow/utils.py:317]: `file_id` is not a URL.

File: src/pystow/utils.py [hunk 4]
```diff
@@ -306,35 +305,31 @@ def download_from_google(
...
     except (Exception, KeyboardInterrupt):
         if clean_on_failure:
-            try:
-                os.remove(path)
-            except OSError:
-                pass  # if the file can't be deleted then no problem
+            path.unlink(missing_ok=True)
         raise
 
-    raise_on_digest_mismatch(path=path, hexdigests=hexdigests)
```
Comment (mberr) [inline src/pystow/utils.py]: no integrity check after download?

Comment (cthoyt) [reply src/pystow/utils.py]: oops I did a bad job with merging from master

File: src/pystow/utils.py (hunk 3 above)
Comment (cthoyt) [reply src/pystow/utils.py:320]: oh, I accidentally clobbered your updated text here too. soz

**Review by cthoyt — APPROVED** [review]

**Pull request merged**